bookkeeper-server/src/test/java/org/apache/bookkeeper/meta/ZkLedgerUnderreplicationManagerTest.java

```diff
@@ -0,0 +1,244 @@
+    }
+
+    @Test
+    public void testMarkLedgerUnderreplicatedConcurrently() throws Exception {
+        final int numLedgers = 20;
+        List<CompletableFuture<Void>> futures = Lists.newArrayListWithExpectedSize(numLedgers);
+        long ledgerId = 0xabcc00;
+        Set<String> expectedBookies = Sets.newHashSet();
+        for (int i = 0; i < numLedgers; i++) {
+            futures.add(
+                urMgr.markLedgerUnderreplicatedAsync(
+                    ledgerId, Lists.newArrayList("bookie-" + i)));
+            expectedBookies.add("bookie-" + i);
+        }
+        FutureUtils.result(FutureUtils.collect(futures));
+
+        UnderreplicatedLedgerFormat urLedgerFormat = urMgr.getLedgerUnreplicationInfo(ledgerId);
+        Set<String> actualBookies = Sets.newHashSet();
+        actualBookies.addAll(urLedgerFormat.getReplicaList());
+
+        assertEquals(expectedBookies, actualBookies);
+    }
+
+    @Test
+    public void testMarkLedgerUnderreplicatedConcurrentlyWithDifferentClients() throws Exception {
+        final int numLedgers = 20;
+        List<ZooKeeper> zks = new ArrayList<>(numLedgers);
+        List<ZkLedgerUnderreplicationManager> urMgrs = new ArrayList<>(numLedgers);
+
+        for (int i = 0; i < numLedgers; i++) {
+            zks.add(createZooKeeper());
+            urMgrs.add(new ZkLedgerUnderreplicationManager(baseServerConf, zks.get(i)));
+        }
+
+        List<CompletableFuture<Void>> futures = Lists.newArrayListWithExpectedSize(numLedgers);
+        long ledgerId = 0xabcd00;
+        Set<String> expectedBookies = Sets.newHashSet();
+        for (int i = 0; i < numLedgers; i++) {
+            futures.add(
+                urMgrs.get(i).markLedgerUnderreplicatedAsync(
+                    ledgerId, Lists.newArrayList("bookie-" + i)));
+            expectedBookies.add("bookie-" + i);
+        }
+
+        FutureUtils.result(FutureUtils.collect(futures));
+
+        UnderreplicatedLedgerFormat urLedgerFormat = urMgr.getLedgerUnreplicationInfo(ledgerId);
+        Set<String> actualBookies = Sets.newHashSet();
+        actualBookies.addAll(urLedgerFormat.getReplicaList());
+
+        assertEquals(expectedBookies, actualBookies);
+
+        for (LedgerUnderreplicationManager urMgr : urMgrs) {
+            urMgr.close();
+        }
+        for (ZooKeeper zk : zks) {
+            zk.close();
+        }
+    }
+
+
+    @Test
+    public void testMarkLedgerUnderreplicatedWhenSessionExpired() throws Exception {
+        final long ledgerId = 0xabbd00;
+        @Cleanup
+        ZooKeeper zk = new ZooKeeper(getZooKeeperConnectString(), 60000, null);
+        @Cleanup
+        LedgerUnderreplicationManager urMgr = new ZkLedgerUnderreplicationManager(baseServerConf, zk);
+        // open another zookeeper client to expire current session
+        @Cleanup
+        ZooKeeper newZk = new ZooKeeper(
+            getZooKeeperConnectString(), 60000, null, zk.getSessionId(), zk.getSessionPasswd());
```

Comment:
hmm..forgot the reason, for why would it cause the existing zkc session to expire